apps/lib/apps.c

```diff
@@ -2288,16 +2288,14 @@ int cert_matches_key(const X509 *cert, const EVP_PKEY *pkey)
 }
 
 /* Ensure RFC 5280 compliance, adapt keyIDs as needed, and sign the cert info */
-int do_X509_sign(X509 *cert, EVP_PKEY *pkey, const char *md,
+int do_X509_sign(X509 *cert, int force_v1, EVP_PKEY *pkey, const char *md,
                  STACK_OF(OPENSSL_STRING) *sigopts, X509V3_CTX *ext_ctx)
 {
-    const STACK_OF(X509_EXTENSION) *exts = X509_get0_extensions(cert);
     EVP_MD_CTX *mctx = EVP_MD_CTX_new();
     int self_sign;
     int rv = 0;
 
-    if (sk_X509_EXTENSION_num(exts /* may be NULL */) > 0) {
-        /* Prevent X509_V_ERR_EXTENSIONS_REQUIRE_VERSION_3 */
+    if (!force_v1) {
```

Comment:
Should you also check that sk_X509_EXTENSION_num(exts) == 0 if force_v1 is set?

Reply:
I thought about this, too.
Allowing for `sk_X509_EXTENSION_num(exts) > 0` to override `force_v1` will require tuning down the name and description of `force_v1` and the `-x509v1` option.
And I believe this should be more generally done at libcrypto level., as I suggested above:
Whenever in the end just before signing a cert there are some extensions, promote the cert to V3.

Reply:
Shall I weaken `force_v1` like this
and do the suggested change in `X509_sign{_ctx}()` (to set version 3 if extensions are present)?
This would make sure that certs bear version 3 whenever needed.

Reply:
Done.